diff -u -r gcc-4.1.0-orig/gcc/config/rs6000/rs6000.c gcc-4.1.0/gcc/config/rs6000/rs6000.c
--- gcc-4.1.0-orig/gcc/config/rs6000/rs6000.c	2006-02-10 12:58:33.000000000 -0800
+++ gcc-4.1.0/gcc/config/rs6000/rs6000.c	2025-03-31 01:11:52.702106712 -0700
@@ -3565,7 +3565,7 @@
 
   /* The TOC register is not killed across calls in a way that is
      visible to the compiler.  */
-  if (DEFAULT_ABI == ABI_AIX)
+  if ((DEFAULT_ABI == ABI_AIX) || (target_flags & MASK_WINNT))
     call_really_used_regs[2] = 0;
 
   if (DEFAULT_ABI == ABI_V4
@@ -3992,7 +3992,7 @@
 
 	  /* If this is a function address on -mcall-aixdesc,
 	     convert it to the address of the descriptor.  */
-	  if (DEFAULT_ABI == ABI_AIX
+	  if (((DEFAULT_ABI == ABI_AIX) || (target_flags & MASK_WINNT))
 	      && GET_CODE (operands[1]) == SYMBOL_REF
 	      && XSTR (operands[1], 0)[0] == '.')
 	    {
@@ -10081,6 +10081,7 @@
 	  break;
 
 	case ABI_V4:
+    break;
 	case ABI_DARWIN:
 	  break;
 	}
@@ -10629,10 +10630,15 @@
 #endif
 	  assemble_name (file, name);
 	}
+      if (target_flags & MASK_WINNT)
+        {
+          fputs("..", file);
+          rs6000_output_function_entry (file, XSTR (x, 0));
+        }
       else if (!DOT_SYMBOLS)
-	assemble_name (file, XSTR (x, 0));
+        assemble_name (file, XSTR (x, 0));
       else
-	rs6000_output_function_entry (file, XSTR (x, 0));
+        rs6000_output_function_entry (file, XSTR (x, 0));
       return;
 
     case 'Z':
@@ -10870,8 +10876,8 @@
 	 address into the address of the descriptor, not the function
 	 itself.  */
       else if (GET_CODE (x) == SYMBOL_REF
-	       && XSTR (x, 0)[0] == '.'
-	       && DEFAULT_ABI == ABI_AIX)
+               && XSTR (x, 0)[0] == '.'
+               && ((DEFAULT_ABI == ABI_AIX) || (target_flags & MASK_WINNT)))
 	{
 	  const char *name = XSTR (x, 0);
 	  while (*name == '.')
@@ -17183,8 +17189,9 @@
   enum machine_mode pmode = Pmode;
   int regsize = (TARGET_32BIT) ? 4 : 8;
   rtx ctx_reg = force_reg (pmode, cxt);
+  int is_winnt = (target_flags & MASK_WINNT) ? 0x80000000 : 0;
 
-  switch (DEFAULT_ABI)
+  switch (is_winnt ? is_winnt : DEFAULT_ABI)
     {
     default:
       gcc_unreachable ();
@@ -17197,17 +17204,18 @@
     /* Under AIX, just build the 3 word function descriptor */
     case ABI_AIX:
       {
-	rtx fn_reg = gen_reg_rtx (pmode);
-	rtx toc_reg = gen_reg_rtx (pmode);
-	emit_move_insn (fn_reg, MEM_DEREF (fnaddr));
-	emit_move_insn (toc_reg, MEM_PLUS (fnaddr, regsize));
-	emit_move_insn (MEM_DEREF (addr), fn_reg);
-	emit_move_insn (MEM_PLUS (addr, regsize), toc_reg);
-	emit_move_insn (MEM_PLUS (addr, 2*regsize), ctx_reg);
+        rtx fn_reg = gen_reg_rtx (pmode);
+        rtx toc_reg = gen_reg_rtx (pmode);
+        emit_move_insn (fn_reg, MEM_DEREF (fnaddr));
+        emit_move_insn (toc_reg, MEM_PLUS (fnaddr, regsize));
+        emit_move_insn (MEM_DEREF (addr), fn_reg);
+        emit_move_insn (MEM_PLUS (addr, regsize), toc_reg);
+        emit_move_insn (MEM_PLUS (addr, 2*regsize), ctx_reg);
       }
       break;
 
     /* Under V.4/eabi/darwin, __trampoline_setup does the real work.  */
+    case 0x80000000:
     case ABI_DARWIN:
     case ABI_V4:
       emit_library_call (gen_rtx_SYMBOL_REF (SImode, "__trampoline_setup"),
@@ -18077,27 +18085,39 @@
   ASM_OUTPUT_TYPE_DIRECTIVE (file, name, "function");
   ASM_DECLARE_RESULT (file, DECL_RESULT (decl));
 
-  if (DEFAULT_ABI == ABI_AIX)
+  if ((DEFAULT_ABI == ABI_AIX) || (target_flags & MASK_WINNT))
     {
       const char *desc_name, *orig_name;
 
       orig_name = (*targetm.strip_name_encoding) (name);
       desc_name = orig_name;
       while (*desc_name == '.')
-	desc_name++;
+        desc_name++;
 
-      if (TREE_PUBLIC (decl))
-	fprintf (file, "\t.globl %s\n", desc_name);
+      if (TREE_PUBLIC (decl)) {
+        fprintf (file, "\t.globl %s\n", desc_name);
+        if (target_flags & MASK_WINNT) {
+          fprintf (file, "\t.globl ..%s\n", desc_name);
+        }
+      }
 
       fprintf (file, "%s\n", MINIMAL_TOC_SECTION_ASM_OP);
       fprintf (file, "%s:\n", desc_name);
-      fprintf (file, "\t.long %s\n", orig_name);
+      fprintf (file, "\t.long ..%s\n", orig_name);
       fputs ("\t.long _GLOBAL_OFFSET_TABLE_\n", file);
       if (DEFAULT_ABI == ABI_AIX)
-	fputs ("\t.long 0\n", file);
-      fprintf (file, "\t.previous\n");
+        fputs ("\t.long 0\n", file);
+      if (target_flags & MASK_WINNT) {
+        fprintf (file, "\t.text\n");
+      } else {
+        fprintf (file, "\t.previous\n");
+      }
     }
-  ASM_OUTPUT_LABEL (file, name);
+  if (target_flags & MASK_WINNT) {
+    fprintf(file, "..%s:\n", name);
+  } else {
+    ASM_OUTPUT_LABEL (file, name);
+  }
 }
 
 static void
diff -u -r gcc-4.1.0-orig/gcc/config/rs6000/rs6000.h gcc-4.1.0/gcc/config/rs6000/rs6000.h
--- gcc-4.1.0-orig/gcc/config/rs6000/rs6000.h	2006-02-10 12:58:33.000000000 -0800
+++ gcc-4.1.0/gcc/config/rs6000/rs6000.h	2025-03-31 00:01:36.727533592 -0700
@@ -1237,9 +1237,11 @@
   ABI_NONE,
   ABI_AIX,			/* IBM's AIX */
   ABI_V4,			/* System V.4/eabi */
-  ABI_DARWIN			/* Apple's Darwin (OS X kernel) */
+  ABI_DARWIN,			/* Apple's Darwin (OS X kernel) */
 };
 
+#define ABI_NT ABI_V4
+
 extern enum rs6000_abi rs6000_current_abi;	/* available for use by subtarget */
 
 /* Define this if pushing a word on the stack
@@ -3072,4 +3074,3 @@
 
 extern GTY(()) tree rs6000_builtin_types[RS6000_BTI_MAX];
 extern GTY(()) tree rs6000_builtin_decls[RS6000_BUILTIN_COUNT];
-
diff -u -r gcc-4.1.0-orig/gcc/config/rs6000/rs6000.md gcc-4.1.0/gcc/config/rs6000/rs6000.md
--- gcc-4.1.0-orig/gcc/config/rs6000/rs6000.md	2006-02-19 11:31:22.000000000 -0800
+++ gcc-4.1.0/gcc/config/rs6000/rs6000.md	2025-03-31 02:29:30.553958110 -0700
@@ -9951,19 +9951,6 @@
 ;; variable argument function.  It is > 0 if FP registers were passed
 ;; and < 0 if they were not.
 
-(define_insn "*call_indirect_nonlocal_aix32"
-  [(call (mem:SI (match_operand:SI 0 "register_operand" "c,*l"))
-	 (match_operand 1 "" "g,g"))
-   (use (reg:SI 2))
-   (use (reg:SI 11))
-   (set (reg:SI 2)
-	(mem:SI (plus:SI (reg:SI 1) (const_int 20))))
-   (clobber (match_scratch:SI 2 "=l,l"))]
-  "TARGET_32BIT && DEFAULT_ABI == ABI_AIX"
-  "b%T0l\;{l|lwz} 2,20(1)"
-  [(set_attr "type" "jmpreg")
-   (set_attr "length" "8")])
-
 (define_insn "*call_nonlocal_aix32"
   [(call (mem:SI (match_operand:SI 0 "symbol_ref_operand" "s"))
 	 (match_operand 1 "" "g"))
@@ -10069,13 +10056,20 @@
   "DEFAULT_ABI == ABI_V4
    || DEFAULT_ABI == ABI_DARWIN"
 {
+  output_asm_insn("stw 31,-16(1)", operands);
+  output_asm_insn("stw 2,-12(1)", operands);
+  output_asm_insn("mfctr 31", operands);
+  output_asm_insn("lwz 2,4(31)", operands);
+  output_asm_insn("lwz 31,0(31)", operands);
+  output_asm_insn("mtctr 31", operands);
+  output_asm_insn("lwz 31,-16(1)", operands);
   if (INTVAL (operands[2]) & CALL_V4_SET_FP_ARGS)
     output_asm_insn ("crxor 6,6,6", operands);
 
   else if (INTVAL (operands[2]) & CALL_V4_CLEAR_FP_ARGS)
     output_asm_insn ("creqv 6,6,6", operands);
 
-  return "b%T0l";
+  return "b%T0l;lwz 2,-12(1)";
 }
   [(set_attr "type" "jmpreg,jmpreg,jmpreg,jmpreg")
    (set_attr "length" "4,4,8,8")])
diff -u -r gcc-4.1.0-orig/gcc/config/rs6000/rs6000.opt gcc-4.1.0/gcc/config/rs6000/rs6000.opt
--- gcc-4.1.0-orig/gcc/config/rs6000/rs6000.opt	2005-11-14 21:00:40.000000000 -0800
+++ gcc-4.1.0/gcc/config/rs6000/rs6000.opt	2025-03-30 19:53:34.256492763 -0700
@@ -36,6 +36,10 @@
 Target Report RejectNegative Mask(POWERPC)
 Use PowerPC instruction set
 
+mwinnt
+Target Report Mask(WINNT)
+Build for Windows NT
+
 mno-powerpc
 Target Report RejectNegative
 Do not use PowerPC instruction set
diff -u -r gcc-4.1.0-orig/gcc/config/rs6000/sysv4.h gcc-4.1.0/gcc/config/rs6000/sysv4.h
--- gcc-4.1.0-orig/gcc/config/rs6000/sysv4.h	2006-02-10 12:58:33.000000000 -0800
+++ gcc-4.1.0/gcc/config/rs6000/sysv4.h	2025-03-30 20:21:10.421436959 -0700
@@ -89,6 +89,11 @@
 									\
   if (!strcmp (rs6000_abi_name, "sysv"))				\
     rs6000_current_abi = ABI_V4;					\
+  else if (!strcmp (rs6000_abi_name, "nt")) \
+    {                                                         \
+  rs6000_current_abi = ABI_V4;                                \
+      target_flags |= MASK_WINNT;                             \
+      }                                                   \
   else if (!strcmp (rs6000_abi_name, "sysv-noeabi"))			\
     {									\
       rs6000_current_abi = ABI_V4;					\
diff -u -r gcc-4.1.0-orig/gcc/toplev.h gcc-4.1.0/gcc/toplev.h
--- gcc-4.1.0-orig/gcc/toplev.h	2005-09-08 17:47:05.000000000 -0700
+++ gcc-4.1.0/gcc/toplev.h	2024-11-05 19:49:36.523930775 -0800
@@ -171,6 +171,7 @@
 #  define CTZ_HWI __builtin_ctz
 # endif
 
+#ifdef NO_PPC_GCC
 extern inline int
 floor_log2 (unsigned HOST_WIDE_INT x)
 {
@@ -182,6 +183,7 @@
 {
   return x == (x & -x) && x ? (int) CTZ_HWI (x) : -1;
 }
+#endif
 #endif /* GCC_VERSION >= 3004 */
 
 /* Functions used to get and set GCC's notion of in what directory
diff -u -r gcc-4.1.0-orig/libssp/gets-chk.c gcc-4.1.0/libssp/gets-chk.c
--- gcc-4.1.0-orig/libssp/gets-chk.c	2005-08-16 20:26:45.000000000 -0700
+++ gcc-4.1.0/libssp/gets-chk.c	2024-11-05 19:10:21.514130939 -0800
@@ -52,6 +52,7 @@
 # include <string.h>
 #endif
 
+#if 0
 extern void __chk_fail (void) __attribute__((__noreturn__));
 
 char *
@@ -86,3 +87,4 @@
     free (buf);
   return ret;
 }
+#endif
diff -u -r gcc-4.1.0-orig/libssp/Makefile.in gcc-4.1.0/libssp/Makefile.in
--- gcc-4.1.0-orig/libssp/Makefile.in	2006-02-27 16:29:00.000000000 -0800
+++ gcc-4.1.0/libssp/Makefile.in	2024-11-05 19:10:21.512130899 -0800
@@ -71,10 +71,11 @@
 	"$(DESTDIR)$(libsubincludedir)"
 toolexeclibLTLIBRARIES_INSTALL = $(INSTALL)
 LTLIBRARIES = $(toolexeclib_LTLIBRARIES)
-am_libssp_la_OBJECTS = ssp.lo gets-chk.lo memcpy-chk.lo memmove-chk.lo \
-	mempcpy-chk.lo memset-chk.lo snprintf-chk.lo sprintf-chk.lo \
-	stpcpy-chk.lo strcat-chk.lo strcpy-chk.lo strncat-chk.lo \
-	strncpy-chk.lo vsnprintf-chk.lo vsprintf-chk.lo
+am_libssp_la_OBJECTS = 
+#am_libssp_la_OBJECTS = ssp.lo gets-chk.lo memcpy-chk.lo memmove-chk.lo \
+#	mempcpy-chk.lo memset-chk.lo snprintf-chk.lo sprintf-chk.lo \
+#	stpcpy-chk.lo strcat-chk.lo strcpy-chk.lo strncat-chk.lo \
+#	strncpy-chk.lo vsnprintf-chk.lo vsprintf-chk.lo
 libssp_la_OBJECTS = $(am_libssp_la_OBJECTS)
 am_libssp_nonshared_la_OBJECTS = libssp_nonshared_la-ssp-local.lo
 libssp_nonshared_la_OBJECTS = $(am_libssp_nonshared_la_OBJECTS)
diff -u -r gcc-4.1.0-orig/libssp/memcpy-chk.c gcc-4.1.0/libssp/memcpy-chk.c
--- gcc-4.1.0-orig/libssp/memcpy-chk.c	2005-08-16 20:26:45.000000000 -0700
+++ gcc-4.1.0/libssp/memcpy-chk.c	2024-11-05 19:10:21.515130959 -0800
@@ -39,6 +39,7 @@
 # include <string.h>
 #endif
 
+#if 0
 extern void __chk_fail (void) __attribute__((__noreturn__));
 
 void *
@@ -49,3 +50,4 @@
     __chk_fail ();
   return memcpy (dest, src, len);
 }
+#endif
diff -u -r gcc-4.1.0-orig/libssp/memset-chk.c gcc-4.1.0/libssp/memset-chk.c
--- gcc-4.1.0-orig/libssp/memset-chk.c	2005-08-16 20:26:45.000000000 -0700
+++ gcc-4.1.0/libssp/memset-chk.c	2024-11-05 19:10:21.516130980 -0800
@@ -41,6 +41,7 @@
 
 extern void __chk_fail (void) __attribute__((__noreturn__));
 
+#if 0
 void *
 __memset_chk (void *dest, int val, size_t len, size_t slen)
 {
@@ -48,3 +49,4 @@
     __chk_fail ();
   return memset (dest, val, len);
 }
+#endif
diff -u -r gcc-4.1.0-orig/libssp/ssp.c gcc-4.1.0/libssp/ssp.c
--- gcc-4.1.0-orig/libssp/ssp.c	2005-08-16 20:26:45.000000000 -0700
+++ gcc-4.1.0/libssp/ssp.c	2024-11-05 19:10:21.516130980 -0800
@@ -58,6 +58,8 @@
 
 void *__stack_chk_guard = 0;
 
+#if 0
+
 static void __attribute__ ((constructor))
 __guard_setup (void)
 {
@@ -176,3 +178,4 @@
   __stack_chk_fail ();
 }
 #endif
+#endif
diff -u -r gcc-4.1.0-orig/libssp/stpcpy-chk.c gcc-4.1.0/libssp/stpcpy-chk.c
--- gcc-4.1.0-orig/libssp/stpcpy-chk.c	2005-08-16 20:26:45.000000000 -0700
+++ gcc-4.1.0/libssp/stpcpy-chk.c	2024-11-05 19:10:21.517131000 -0800
@@ -41,6 +41,7 @@
 
 extern void __chk_fail (void) __attribute__((__noreturn__));
 
+#if 0
 char *
 __stpcpy_chk (char *__restrict__ dest, const char *__restrict__ src,
               size_t slen)
@@ -50,3 +51,4 @@
     __chk_fail ();
   return memcpy (dest, src, len + 1) + len;
 }
+#endif
diff -u -r gcc-4.1.0-orig/libssp/strcat-chk.c gcc-4.1.0/libssp/strcat-chk.c
--- gcc-4.1.0-orig/libssp/strcat-chk.c	2005-08-16 20:26:45.000000000 -0700
+++ gcc-4.1.0/libssp/strcat-chk.c	2024-11-05 19:10:21.517131000 -0800
@@ -41,6 +41,7 @@
 
 extern void __chk_fail (void) __attribute__((__noreturn__));
 
+#if 0
 char *
 __strcat_chk (char *__restrict__ dest, const char *__restrict__ src,
               size_t slen)
@@ -71,3 +72,4 @@
 
   return dest;
 }
+#endif
diff -u -r gcc-4.1.0-orig/libssp/strcpy-chk.c gcc-4.1.0/libssp/strcpy-chk.c
--- gcc-4.1.0-orig/libssp/strcpy-chk.c	2005-08-16 20:26:45.000000000 -0700
+++ gcc-4.1.0/libssp/strcpy-chk.c	2024-11-05 19:10:21.518131020 -0800
@@ -39,6 +39,7 @@
 # include <string.h>
 #endif
 
+#if 0
 extern void __chk_fail (void) __attribute__((__noreturn__));
 
 char *
@@ -50,3 +51,4 @@
     __chk_fail ();
   return memcpy (dest, src, len + 1);
 }
+#endif
diff -u -r gcc-4.1.0-orig/libssp/strncat-chk.c gcc-4.1.0/libssp/strncat-chk.c
--- gcc-4.1.0-orig/libssp/strncat-chk.c	2005-08-16 20:26:45.000000000 -0700
+++ gcc-4.1.0/libssp/strncat-chk.c	2024-11-05 19:10:21.518131020 -0800
@@ -41,6 +41,7 @@
 
 extern void __chk_fail (void) __attribute__((__noreturn__));
 
+#if 0
 char *
 __strncat_chk (char *__restrict__ dest, const char *__restrict__ src,
                size_t n, size_t slen)
@@ -118,3 +119,4 @@
 
   return s;
 }
+#endif
diff -u -r gcc-4.1.0-orig/libssp/strncpy-chk.c gcc-4.1.0/libssp/strncpy-chk.c
--- gcc-4.1.0-orig/libssp/strncpy-chk.c	2005-08-16 20:26:45.000000000 -0700
+++ gcc-4.1.0/libssp/strncpy-chk.c	2024-11-05 19:10:21.519131040 -0800
@@ -41,6 +41,8 @@
 
 extern void __chk_fail (void) __attribute__((__noreturn__));
 
+#if 0
+
 #ifdef HAVE_STRNCPY
 char *
 __strncpy_chk (char *__restrict__ dest, const char *__restrict__ src,
@@ -51,3 +53,5 @@
   return strncpy (dest, src, len);
 }
 #endif
+
+#endif
diff -u -r gcc-4.1.0-orig/Makefile.in gcc-4.1.0/Makefile.in
--- gcc-4.1.0-orig/Makefile.in	2005-12-15 06:02:02.000000000 -0800
+++ gcc-4.1.0/Makefile.in	2024-11-05 19:10:21.509130838 -0800
@@ -1,4 +1,4 @@
-
+#
 # Makefile.in is generated from Makefile.tpl by 'autogen Makefile.def'.
 #
 # Makefile for directory with subdirs to build.
@@ -726,6 +726,7 @@
     maybe-all-utils \
     maybe-all-gnattools
 .PHONY: all-target
+#    maybe-all-target-libiberty
 all-target:  \
     maybe-all-target-libstdc++-v3 \
     maybe-all-target-libmudflap \
@@ -2108,6 +2109,7 @@
     maybe-install-gnattools
 
 .PHONY: install-target
+#     maybe-install-target-libiberty
 install-target:  \
     maybe-install-target-libstdc++-v3 \
     maybe-install-target-libmudflap \
@@ -2118,7 +2120,6 @@
     maybe-install-target-libtermcap \
     maybe-install-target-winsup \
     maybe-install-target-libgloss \
-    maybe-install-target-libiberty \
     maybe-install-target-gperf \
     maybe-install-target-examples \
     maybe-install-target-libffi \
